update post: Managing Multiple Okta Instances with Terraform Cloud (#958)
oktadeveloper/okta-blog: _source/_posts/2020-02-03-managing-multiple-okta-instances-with-terraform-cloud.md

```console
git init
git add identity.tf okta.auto.tfvars
git commit -m "Initial commit. Adding the date of birth extension"
```

> **NOTE**: Don't push your `secret.auto.tfvars` file that contains the `api_token` to source control. Add it into your `.gitignore` file, so you don't accidentally include it. You can also ignore `.tfstate` and `.hcl` files and the `.terraform` folder.

Now that you've defined your Terraform rules, you can apply these changes by running `terraform apply` on the command line. If you do this, Terraform will actually modify your user schema as defined above by talking to the Okta APIs. It will also update the state file on your local machine and keep track of the resources it creates.

## Get Started with Terraform Cloud

> Although working on your local machine is fine if you're building a simple project solo, in many cases, you'll want to use [Terraform Cloud](https://www.terraform.io/docs/cloud/index.html), which allows you have a nice Terraform workflow without needing to track state files locally. It's incredibly convenient to use for any _real_ project.

Reviewer: ```suggestion
Although working on your local machine is fine if you're building a simple project solo, in many cases, you'll want to use [Terraform Cloud](https://www.terraform.io/docs/cloud/index.html), which allows you to have a nice Terraform workflow without needing to track state files locally. It's incredibly convenient to use for any _real_ project.
```

Now let's set up Terraform Cloud so you can configure your infrastructure without needing to worry about storing and managing local files.

If you don't already have a Terraform Cloud account, you can sign up for a (free account)[https://app.terraform.io/signup/account], which you can use with a team of up to five people.

Once you create an account and sign in, you must choose the **Start from scratch** option on the welcome screen. On the next screen, create an organization and proceed. Now, you are going to create a "Workspace". Workspaces describe your environments (production, staging, development, etc.). The environment you'll create below will be a "production" environment (we'll create others later).

{% img blog/okta-terraform-cloud/terraform-cloud-create-workspace.png alt:"Create workspace using Terraform Cloud" width:"800" %}{: .center-image }

> On the first page of the Workspace creation flow, select **Version control flow**, as we will be saving our configuration in a Git repository. Next, you need to select your source control provider. In my case, I'm using GitHub to store my code, so I'll select GitHub and grant permission for Terraform Cloud to access my project. When GitHub prompts to select repositories to install Terraform cloud, choose one or more repositories where you intent to store Terraform configurations.

Reviewer 2: ```suggestion
On the first page of the Workspace creation flow, select **Version control flow**, as we will be saving our configuration in a Git repository. Next, you need to select your source control provider. In my case, I'm using GitHub to store my code, so I'll select GitHub and grant permission for Terraform Cloud to access my project. When GitHub prompts to select repositories to install Terraform cloud, choose one or more repositories where you intend to store Terraform configurations.
```
typo

{% img blog/okta-terraform-cloud/terraform-cloud-select-repository.png alt:"Terraform Cloud select provider" width:"800" %}{: .center-image }

After authorizing with GitHub, click the GitHub button again to choose the repository that holds the project we created. If you haven't already, push the `okta-user-schema` project to this repository.

{% img blog/okta-terraform-cloud/terraform-cloud-select-repository-2.png alt:"Terraform Cloud select repository" width:"800" %}{: .center-image }

Give your new workspace a descriptive name. We're defining our production environment in this example, so use the repository's name with `-production` added to the end for clarity. For example, `my-project-production`.

{% img blog/okta-terraform-cloud/terraform-cloud-name-workspace.png alt:"Terraform Cloud name workspace" width:"800" %}{: .center-image }

> Locally you defined variables in the `okta.auto.tfvars` file. But we didn't add the `api_token` to this and hence isn't stored in source control; you have to define the variable not committed to source control explicitly. Click the **Variables** tab in the top navbar of the workspace, then define the `api_token` variable.

Reviewer: Nit. I had to read this sentence a couple of times to understand. But I also haven't worked with Terraform before. How about 

> Locally you defined variables in the `okta.auto.tfvars` file. But we didn't add the `api-token` variable to this file in order to exclude it from source control. Therefore, you have to define the variable explicitly.

> Locally you defined variables in the `okta.auto.tfvars` file. But we didn't add the `api_token` to this and hence isn't stored in source control; you have to define the variable not committed to source control explicitly. Click the **Variables** tab in the top navbar of the workspace, then define the `api_token` variable.

Author: ```suggestion
Locally you defined variables in the `okta.auto.tfvars` file. But we didn't add the `api_token` variable to that file for obvious security reasons. Hence that variable is not stored in source control; Therefore, you have to define the variable explicitly in the Terraform workspace. Click the **Variables** tab in the top navbar of the workspace, then define the `api_token` variable.
```

> Locally you defined variables in the `okta.auto.tfvars` file. But we didn't add the `api_token` variable to that file for obvious security reasons. Hence that variable is not stored in source control; Therefore, you have to define the variable explicitly in the Terraform workspace. Click the **Variables** tab in the top navbar of the workspace, then define the `api_token` variable.

Reviewer: Some grammar issues here with the word grouping before the semicolon and capitalization.

Should it be something like one of the following?
> But we didn't add the `api_token` variable to that file for obvious security reasons; hence the variable is not stored in source control. Therefore, you have to define...

or maybe
> But we didn't add the `api_token` variable to that file for obvious security reasons. Since the variable is not in source contr, you have to define...

Author: I'm not a native English speaker so I normally just go with whatever Grammarly suggests, so in this case, I'll go with your first suggestion :smiley_cat:

Reviewer: Grammarly is slacking! Thanks for making the change. 🤸

## Manage Multiple Okta Environments with Terraform

> Now that you're using Terraform Cloud to manage a single environment (production). Let's add a second environment (development), which will make it easy for your developers to run tests against this new Okta environment without impacting production.

Reviewer: The first sentence is a fragment. How about

> Now that you're using Terraform Cloud to manage a single environment (production), let's add a second environment (development). Doing so will make it easy for your developers to run tests against this new Okta environment without impacting production.

> Now that you're using Terraform Cloud to manage a single environment (production). Let's add a second environment (development), which will make it easy for your developers to run tests against this new Okta environment without impacting production.

Author: ```suggestion
Now that you're using Terraform Cloud to manage a single environment (production), let's add a second environment (development). Doing so will make it easy for your developers to run tests against this new Okta environment without impacting production.
```

If you don't have a second Okta org yet, [go create one](https://developer.okta.com/signup/), and then go through the steps to create an API token once more.

Next, create a new Git branch in your project called `dev`. This is where you'll work on your product before releasing it to production (the `main` branch).

```console
git checkout -b dev
git push origin dev
```

In Terraform Cloud, create a new workspace to represent this new environment. Target the same repository as before but set the branch specifier to `dev`.

{% img blog/okta-terraform-cloud/terraform-cloud-create-workspace.jpg alt:"Terraform Cloud create workspace" width:"800" %}{: .center-image }

For this environment, set the apply method in Terraform Cloud to `auto` so changes will be applied to the environment immediately if the planning stage is successful without you needing to confirm anything manually.

{% img blog/okta-terraform-cloud/terraform-cloud-create-workspace-2.jpg alt:"Terraform Cloud workspace settings" width:"800" %}{: .center-image }

> Update the `okta.auto.tfvar` file using your _new_ Okta org's `org_name` and `base_url` and create a new variable for `api_token` from this org. Click **Start new plan** and just like that, Terraform replicated your production Okta configuration from your first Okta org into the second Okta org!

Reviewer: ```suggestion
Update the `okta.auto.tfvar` file using your _new_ Okta org's `org_name` and `base_url` and create a new variable for `api_token` for this org. Click **Start new plan** and just like that, Terraform replicated your production Okta configuration from your first Okta org into the second Okta org!
```

Reviewer: Is this correct? should it be "for"?

Author: Not exactly but I see that the sentence is not clear, I'll rephrase

> Update the `okta.auto.tfvar` file using your _new_ Okta org's `org_name` and `base_url` and create a new variable for `api_token` from this org. Click **Start new plan** and just like that, Terraform replicated your production Okta configuration from your first Okta org into the second Okta org!

Author: ```suggestion
Update the `okta.auto.tfvar` file using your _new_ Okta org's `org_name` and `base_url`. Create a new Terraform Cloud workspace variable for `api_token` and make sure to use the API Token from the new Okta org. Click **Start new plan** and just like that, Terraform replicated your production Okta configuration from your first Okta org into the second Okta org!
```

> Update the `okta.auto.tfvar` file using your _new_ Okta org's `org_name` and `base_url`. Create a new Terraform Cloud workspace variable for `api_token` and make sure to use the API Token from the new Okta org. Click **Start new plan** and just like that, Terraform replicated your production Okta configuration from your first Okta org into the second Okta org!

Reviewer 2: ```suggestion
Update the `okta.auto.tfvar` file using your _new_ Okta org's `org_name` and `base_url`. Create a new Terraform Cloud workspace variable for `api_token` and make sure to use the API Token from the new Okta org. Click **Start new plan** and just like that, Terraform replicates your production Okta configuration from your first Okta org into the second Okta org!
```
verb tense

## Promoting Changes Using Terraform

> Now that you set up Terraform to manage your development and production environments, let's imagine your development team wants to change how Okta is configured. You'd like them to be able to do this, but you'd also still like to control and review these changes before they make their way into production. What do you do?

Reviewer 2: ```suggestion
Now that you've set up Terraform to manage your development and production environments, let's imagine your development team wants to change how Okta is configured. You'd like them to be able to do this, but you'd also still like to control and review these changes before they make their way into production. What do you do?
```
verb tense tweak

}
```

Now save and push these changes to your remote repository.

Terraform will trigger and run your plan (and apply it if successful) to the dev environment.

If you go to **Directory** > **Profile Editor** > **Okta** > **Profile** in your Okta Admin Console of the development org, you should see that Terraform applied both your schema extensions.

Once you're happy with the changes you have made to the development environment, you'll want to promote them to the production environment. To do this, you'll need to create a pull request from your `dev` branch to the `main` branch that controls production. As you have branch protection enabled, this provides an opportunity for others to review the changes and for any discussion to take place before you change the production configuration.

{% img blog/okta-terraform-cloud/github-open-pull-request.png alt:"GitHub open pull request" width:"800" %}{: .center-image }

Once you create the pull request, GitHub provides a nice UI experience that shows the branch protection in action.

{% img blog/okta-terraform-cloud/github-review-required.png alt:"GitHub review required" width:"800" %}{: .center-image }

> The status checks run plans against your two Okta instances so the reviewer can verify what the change would do. Clicking **details** on either of these will open Terraform Cloud and show the output of the plan.

Reviewer 2: ```suggestion
The status checks run plans against your two Okta instances so the reviewer can verify what the change will do. Clicking **details** on either of these will open Terraform Cloud and show the output of the plan.
```
tense
